socket/socket.c

```diff
@@ -0,0 +1,427 @@
+{
+	int isnum;
+	lua_Integer d;
+	lua_getfield(L, idx, k);
+	d = lua_tointegerx(L, -1, &isnum);
+	lua_pop(L, 1);
+	if (!isnum)
+		return def;
+
+	return d;
+}
+
+int luasocket(lua_State *L)
+{
+	int err;
+	sock_t sock;
+
+	err = sock_create(socket_tofamily(L, 1), socket_totype(L, 2), 0, &sock);
+
+	if (err < 0)
+		luaL_error(L, "Socket creation error %d", err);
+
+	*((sock_t *) lua_newuserdata(L, sizeof(sock_t))) = sock;
+	luaL_getmetatable(L, LUA_SOCKET);
+	lua_setmetatable(L, -2);
+	return 1;
+}
+
+int luasocket_bind(lua_State *L)
+{
+	int err;
+	struct sockaddr_in addr;
+	sock_t s = *(sock_t *) luaL_checkudata(L, 1, LUA_SOCKET);
+	const char *ip = luaL_checkstring(L, 2);
+	const int port = luaL_checkinteger(L, 3);
+
+	if (port <= 0 || port > 65535)
+		luaL_argerror(L, 2, "Port number out of range");
+
+	addr.sin_family = s->sk->sk_family;
+	addr.sin_port = htons((u_short) port);
+	inet_pton(s->sk->sk_family, ip, &addr.sin_addr.s_addr);
+
+	if ((err = kernel_bind(s, (struct sockaddr *) &addr, sizeof(addr))) < 0)
+		luaL_error(L, "Socket bind error: %d", err);
+
+	return 0;
+}
+int luasocket_listen(lua_State *L)
+{
+	int err;
+	sock_t s = *(sock_t *) luaL_checkudata(L, 1, LUA_SOCKET);
+	const int backlog = luaL_checkinteger(L, 2);
+
+	if (backlog <= 0)
+		luaL_argerror(L, 2, "Backlog number out of range");
+
+	if ((err = kernel_listen(s, backlog)) < 0)
+		luaL_error(L, "Socket listen error: %d", err);
+
+	return 0;
+}
+int luasocket_accept(lua_State *L)
+{
+	int err;
+	sock_t s = *(sock_t *) luaL_checkudata(L, 1, LUA_SOCKET);
+	const int flags = luaL_optnumber(L, 2, 0);
+	sock_t newsock;
+
+	if ((err = kernel_accept(s, &newsock, flags)) < 0)
+		luaL_error(L, "Socket accept error: %d", err);
+
+	*((sock_t *) lua_newuserdata(L, sizeof(sock_t))) = newsock;
+	luaL_getmetatable(L, LUA_SOCKET);
+	lua_setmetatable(L, -2);
+
+	return 1;
+}
+int luasocket_connect(lua_State *L)
+{
+	int err;
+	struct sockaddr_in addr;
+	sock_t s = *(sock_t *) luaL_checkudata(L, 1, LUA_SOCKET);
+	const char *ip = luaL_checkstring(L, 2);
+	const int port = luaL_checkinteger(L, 3);
+	const int flags = luaL_optinteger(L, 4, 0);
+
+	if (port <= 0 || port > 65535)
+		luaL_argerror(L, 2, "Port number out of range");
+
+	addr.sin_family = s->sk->sk_family;
+	addr.sin_port = htons((u_short) port);
+	inet_pton(s->sk->sk_family, ip, &addr.sin_addr.s_addr);
+
+	if ((err = kernel_connect(s, (struct sockaddr *) &addr, sizeof(addr),
+				  flags)) < 0)
+		luaL_error(L, "Socket connect error: %d", err);
+
+	return 0;
+}
+int luasocket_sendmsg(lua_State *L)
+{
+	int i;
+	int err;
+	size_t size;
+	struct msghdr msg;
+	struct kvec vec;
+	struct sockaddr_in addr;
+	sock_t s = *(sock_t *) luaL_checkudata(L, 1, LUA_SOCKET);
+	char *buffer = NULL;
+
+	luaL_checktype(L, 2, LUA_TTABLE);
+
+	if (lua_istable(L, 3)) {
+		size = luaL_len(L, 3);
+		luaL_argcheck(L, size > 0, 3, "data can not be empty");
+	}
+#ifdef LUANTICK_DATA
```

Comment:
let's change to `IS_ENABLED(CONFIG_LUADATA)`
the cpp will evaluate to true for y and m